## APUNTES EXAME ACUE

Ejercicio Taxis

In [1]:
def calcular_probabilidades_colores(prob_colores, prob_acierto, color_identificado):
    """
    Calcula las probabilidades de que el taxi sea de cada color dado que el testigo identificó un color específico.

    :param prob_colores: Diccionario con la probabilidad de cada color de taxi.
    :param prob_acierto: Probabilidad de acierto del testigo.
    :param color_identificado: Color que el testigo identificó.
    :return: Diccionario con las probabilidades de que el taxi sea de cada color.
    """
    prob_fallo = 1 - prob_acierto  # La probabilidad de fallo del testigo

    # Inicializamos un diccionario para las probabilidades condicionales
    prob_testigo = {}
    for color in prob_colores:
        if color == color_identificado:
            prob_testigo[color] = prob_acierto
        else:
            prob_testigo[color] = prob_fallo / (len(prob_colores) - 1)

    # Calculamos la probabilidad total de que el testigo diga que el taxi es del color identificado
    P_T = sum(prob_testigo[color] * prob_colores[color] for color in prob_colores)

    # Calculamos las probabilidades condicionales usando el teorema de Bayes
    prob_condicionales = {}
    for color in prob_colores:
        P_T_given_C = prob_testigo[color]
        P_C = prob_colores[color]
        prob_condicionales[color] = (P_T_given_C * P_C) / P_T

    return prob_condicionales

# Definimos las probabilidades de los colores de los taxis
prob_colores = {
    'Verde': 0.70,
    'Azul': 0.20,
    'Rojo': 0.10
}

# Definimos la probabilidad de acierto del testigo
prob_acierto = 0.80

# Color identificado por el testigo
color_identificado = 'Rojo'

# Calculamos las probabilidades de que el taxi sea de cada color
probabilidades_colores = calcular_probabilidades_colores(prob_colores, prob_acierto, color_identificado)

# Imprimimos los resultados
for color, probabilidad in probabilidades_colores.items():
    print(f"La probabilidad de que el taxi sea {color} dado que el testigo dijo que es {color_identificado} es: {probabilidad:.2f}")

La probabilidad de que el taxi sea Verde dado que el testigo dijo que es Rojo es: 0.41
La probabilidad de que el taxi sea Azul dado que el testigo dijo que es Rojo es: 0.12
La probabilidad de que el taxi sea Rojo dado que el testigo dijo que es Rojo es: 0.47


# Sistemas de recomendacion:

Basado en contenido: “Muestrame másde lo que he escogido o me ha gustado"

Colaborativo: “Dime lo que está gustandoen la comunidad de usuarios"

Híbrido: combinación de varias estrategiasy fuentes de datos

Preferencias del usuario en un sistema de recomendación basado en contenido

Para estimar las preferencias del usuario en un sistema de recomendación basado en contenido, podemos calcular la frecuencia de compra de cada artículo por el usuario.


In [3]:
import pandas as pd
# Ruta al archivo XLSX
file_path = './TT_dataset_compras.xlsx'

# Cargar los datos desde el archivo XLSX
df = pd.read_excel(file_path)
# df = pd.read_csv(file_path_csv)

# Asegurarnos de que la columna índice sea la correcta
df.set_index('ID', inplace=True)

# Mostrar las primeras filas del dataframe para confirmar la carga
print(df.head())

# Calculamos las preferencias del usuario basado en las compras
def calcular_preferencias_usuario(df):
    # Sumar todas las transacciones para obtener el total de compras de cada producto
    total_compras = df.sum()
    
    # Calcular la proporción de cada producto en relación al total de compras del usuario
    preferencias_usuario = total_compras / total_compras.sum()
    
    return preferencias_usuario

# Calculamos las preferencias del usuario
preferencias_usuario = calcular_preferencias_usuario(df)

# Mostramos las preferencias del usuario
print(preferencias_usuario)

    milk  bread  butter  beer  diapers
ID                                    
1      1      1       0     0        0
2      0      0       1     0        0
3      0      0       0     1        1
4      1      1       1     0        0
5      0      1       0     0        0
milk       0.222222
bread      0.333333
butter     0.222222
beer       0.111111
diapers    0.111111
dtype: float64


# Utilidad

La utilidad es una medida de satisfacción o valor que un individuo asigna a diferentes resultados posibles de una decisión. En el marco de la Teoría de la Utilidad, se utilizan varias herramientas y conceptos para estimar y aplicar esta medida

Una forma de calcularlo es la Teoria de Utilidad Esperada, donde la utilidad de una opcion es la suma de las probabilidades de los elementos por su utilidad esperada.

Por ejemplo:

Opción A:
2500 euros con probabilidad 0.33
2400 euros con probabilidad 0.66
0 con probabilidad 0.1 

Opción B:
2400 euros con probabilidad 0.34
0 con probabilidad 0.66 

Utilidad Esperada A =(2500×0.33)+(2400×0.66)+(0×0.1) = 2409
Utilidad Esperada A =(2400×0.34)+(0×0.66) = 816

Dependiendo del problema se puede modificar el calculo, como en el caso del convexo que es mas conservador y se usaria log(2500 + 1)

¿Qué es la Utilidad?:
1. Perspectiva de Marshall:
• Funcionalidad de una alternativa.
• Se estima comparando los atributos de una alternativa con Teoría de la Utilidad
las preferencias del sujeto.
• Concepto de Utilidad estándar.
2. Perspectiva de Bentham:
• Satisfacción que se espera obtener de una alternativa.
• Concepto original de Utilidad.

# Modelo RISP-DM (Cross-Industry Standard Process for Data Mining)
(1) **Fase más importante**: Business Understanding
Justificación: La fase de entendimiento del negocio es crucial porque establece las bases y objetivos del proyecto. Sin una comprensión clara del problema de negocio y de lo que se quiere lograr, es muy probable que el proyecto no entregue el valor esperado o incluso que falle. Esta fase involucra definir los objetivos del negocio, entender las restricciones, identificar las variables clave y formular las preguntas que el análisis de datos intentará responder. Es aquí donde se alinean las expectativas del negocio con las capacidades técnicas del equipo de Data Science.

(2) **Fase que requiere más tiempo de trabajo**: Data Preparation
Justificación: La preparación de datos tiende a ser la fase más laboriosa y que consume más tiempo en un proyecto de Data Science. Este proceso incluye la recolección de datos, limpieza, transformación, y la integración de diferentes fuentes de datos. A menudo, los datos brutos están incompletos, sucios o en formatos no compatibles, lo cual requiere un esfuerzo considerable para preparar un dataset que sea apto para el análisis y modelado. Según estudios y experiencias de la industria, esta fase puede consumir hasta el 80% del tiempo total del proyecto.

Perfiles más idóneos para cada fase del proyecto
1. Business Understanding
Perfil más idóneo: Economista, Analista de Negocios
Justificación: Los economistas y analistas de negocios son ideales para esta fase porque tienen una comprensión profunda de los procesos y objetivos del negocio. Son capaces de traducir problemas de negocio en términos técnicos y analíticos que puedan ser abordados por el equipo de Data Science. Además, tienen la habilidad de identificar oportunidades y restricciones desde una perspectiva estratégica.

2. Data Understanding
Perfil más idóneo: Estadístico, Científico de Datos
Justificación: Los estadísticos y científicos de datos tienen las habilidades necesarias para explorar y analizar los datos de manera inicial. Pueden identificar patrones, tendencias y anomalías, y evaluar la calidad y relevancia de los datos para el proyecto. Su conocimiento en técnicas exploratorias y estadísticas es esencial para esta fase.

3. Data Preparation
Perfil más idóneo: Ingeniero Informático, Científico de Datos
Justificación: Los ingenieros informáticos son expertos en manejar grandes volúmenes de datos, escribir scripts de limpieza, y transformar datos de diferentes fuentes y formatos. Los científicos de datos también juegan un papel importante aquí, ya que entienden los requisitos del modelado y pueden preparar los datos de manera adecuada para los algoritmos de machine learning.

4. Modeling
Perfil más idóneo: Matemático, Estadístico, Científico de Datos
Justificación: Los matemáticos y estadísticos tienen una sólida formación en teorías y métodos analíticos que son fundamentales para construir modelos precisos y robustos. Los científicos de datos, con su conocimiento práctico de los algoritmos de machine learning y herramientas de software, también son esenciales para esta fase.

5. Evaluation
Perfil más idóneo: Estadístico, Científico de Datos
Justificación: La evaluación de los modelos requiere una comprensión profunda de las métricas de desempeño y la capacidad de interpretar los resultados de manera significativa. Los estadísticos y científicos de datos están capacitados para realizar estas evaluaciones, ajustar los modelos y validar su efectividad.

6. Deployment
Perfil más idóneo: Ingeniero Informático, Científico de Datos
Justificación: El despliegue de modelos en producción necesita conocimientos de ingeniería de software, infraestructuras de TI y DevOps. Los ingenieros informáticos son esenciales para esta fase debido a su experiencia en estos campos. Los científicos de datos también contribuyen asegurando que los modelos se integren correctamente y funcionen como se espera en un entorno real.